### PR TITLE
Google Fonts: enqueue only picked font families in Gutenberg

### PR DESCRIPTION
**Pull request by zaguiini**
#### Changes proposed in this Pull Request:

As https://github.com/WordPress/gutenberg/pull/39559 got merged, we have more control over which font faces get to the front-end after registering.

This PR introduces code that introspects block usage and global styles configuration looking for Google fonts that have been registered.

#### Jetpack product discussion

N/A.

#### Does this pull request change what data or activity we track or use?

No.

#### Testing instructions:

Setting up the environment:

1. Run `trunk` Gutenberg;
2. Map Gutenberg to Jetpack docker volume (doing this instead of `a8c-wp-env` because we had loads of trouble trying to get it working);
3. Run `composer install` inside `projects/plugins/jetpack` so the module gets linked with the plugin;
4. Enable the Google fonts module in Jetpack settings;

Enqueueing fonts used in blocks:

1. Open the site editor;
2. Include a block that allows typography customization;
3. Pick one of the registered Google fonts (e.g. Roboto);
4. Save the changes;
5. Open the site.

In the page source, check that only the picked font family shows up in the `style#webfonts-inline-css` element.

Enqueueing fonts used in global styles:

1. Open the site editor;
2. Open the global styles settings;
3. Pick one of the registered Google fonts (e.g. Roboto) in the typography section;
4. Save the changes;
5. Open the site.

In the page source, check that only the picked font families, both used in blocks above and the one that you just set in global styles, shows up in the `style#webfonts-inline-css` element.

### CODE REVIEW

File: projects/plugins/jetpack/modules/google-fonts/class-jetpack-google-fonts.php [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+		add_action( 'after_setup_theme', array( $this, 'register_google_fonts' ) );
+
+		/**
+		 * We are already enqueueing all registered fonts by default when loading the block editor,
+		 * so we only need to scan for webfonts when browsing as a guest.
+		 */
+		if ( ! is_admin() ) {
+			add_action( 'wp_loaded', array( $this, 'scan_global_styles_for_google_fonts' ) );
+			add_filter( 'pre_render_block', array( $this, 'scan_block_for_google_fonts' ), 10, 2 );
+		}
```
Comment (zaguiini) [inline projects/plugins/jetpack/modules/google-fonts/class-jetpack-google-fonts.php:32]: Need to keep it like that until Gutenberg lazy loads font-face declarations as they're picked in the block editor.

File: projects/plugins/jetpack/modules/google-fonts/class-jetpack-google-fonts.php [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+	 *
+	 * @param string $font_family_slug The block content.
+	 */
+	private function maybe_enqueue_font_family( $font_family_slug ) {
+		if ( ! isset( $this->registered_google_fonts[ $font_family_slug ] ) ) {
+			// Font not allow-listed.
+			return;
+		}
+
+		wp_enqueue_webfont( $font_family_slug );
```
Comment (zaguiini) [inline projects/plugins/jetpack/modules/google-fonts/class-jetpack-google-fonts.php:124]: This will trigger a warning if the font is already registered. https://github.com/WordPress/gutenberg/pull/40060 will make the warning disappear.

File: projects/plugins/jetpack/modules/google-fonts/class-jetpack-google-fonts.php [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+	}
+
+	/**
+	 * Scan global styles for Google fonts.
+	 */
+	public function scan_global_styles_for_google_fonts() {
```
Comment (zaguiini) [inline projects/plugins/jetpack/modules/google-fonts/class-jetpack-google-fonts.php:130]: This function probably does not belong here. Right now, however, I'm struggling to find a better place to put it...

When it's time, we can probably move this to Gutenberg.

File: projects/plugins/jetpack/modules/google-fonts/class-jetpack-google-fonts.php [hunk 4]
```diff
@@ -0,0 +1,193 @@
+<?php // phpcs:ignore WordPress.Files.FileName.InvalidClassFileName
+/**
+ * Module: Jetpack Google Fonts
+ *
+ * @package automattic/jetpack
+ */
+class JetpackGoogleFonts {
```
Comment (jeherve) [inline projects/plugins/jetpack/modules/google-fonts/class-jetpack-google-fonts.php:7]: Maybe we could avoid the phpcs issue with a different naming?

```suggestion
<?php
/**
 * Utilities to enable our Google Fonts when the module is enabled.
 *
 * @package automattic/jetpack
 */

namespace Automattic\Jetpack;

/**
 * Register Jetpack Google Fonts and add support for it.
 */
class Google_Fonts {
```

File: projects/plugins/jetpack/modules/google-fonts/jetpack-google-fonts.php [hunk 5]
```diff
@@ -0,0 +1,48 @@
+<?php
```
Comment (jeherve) [inline projects/plugins/jetpack/modules/google-fonts/jetpack-google-fonts.php:1]: I was curious if you had a specific reason to move this to a new file instead of leaving it in `modules/google-fonts.php`? I figure on the long term, the list of fonts may be what we want to edit the most often, and the first place folks will look for it will be `modules/google-fonts.php`, the entry point? What do you think?

File: projects/plugins/jetpack/modules/google-fonts/class-jetpack-google-fonts.php (hunk 3 above)
Comment (jeherve) [reply projects/plugins/jetpack/modules/google-fonts/class-jetpack-google-fonts.php:130]: I wonder if this could be in the package for now, maybe in the Utils class?
https://github.com/Automattic/jetpack/tree/212b4d7bc0ab2be67b0ed7347f6496454e390052/projects/packages/google-fonts-provider/src

File: projects/plugins/jetpack/modules/google-fonts/jetpack-google-fonts.php [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+	'Rubik',
+	'Source Sans Pro',
+	'Source Serif Pro',
+	'Space Mono',
+	'Work Sans',
+);
+
+require __DIR__ . '/class-jetpack-google-fonts.php';
+
+new JetpackGoogleFonts();
```
Comment (jeherve) [inline projects/plugins/jetpack/modules/google-fonts/jetpack-google-fonts.php:46]: That will need to change if we update the class name as I suggested above.

**Review by jeherve — COMMENTED** [review]
I left a few comments below, let me know what you think!

Also, could you add a changelog entry to your PR, using `jetpack changelog add plugins/jetpack`?

Thank you!

File: projects/plugins/jetpack/modules/google-fonts/class-jetpack-google-fonts.php [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+		 * @module google-fonts
+		 *
+		 * @since 10.8
+		 *
+		 * @param array $fonts_to_register Array of Google Font names to register.
+		 */
+		$fonts_to_register = apply_filters( 'jetpack_google_fonts_list', JETPACK_GOOGLE_FONTS_LIST );
+
+		foreach ( $fonts_to_register as $font_family ) {
+			$font_family_slugs = wp_register_webfonts(
```
Comment (zaguiini) [inline projects/plugins/jetpack/modules/google-fonts/class-jetpack-google-fonts.php:66]: This change is included in https://github.com/WordPress/gutenberg/pull/40120.

File: projects/plugins/jetpack/modules/google-fonts/class-jetpack-google-fonts.php (hunk 3 above)
Comment (zaguiini) [reply projects/plugins/jetpack/modules/google-fonts/class-jetpack-google-fonts.php]: I actually created a brand new package. It didn't feel connected to the provider at all, and it could be used by other modules so hurray for reusability!

File: projects/plugins/jetpack/modules/google-fonts/jetpack-google-fonts.php (hunk 5 above)
Comment (zaguiini) [reply projects/plugins/jetpack/modules/google-fonts/jetpack-google-fonts.php:1]: Sure! Moving it to the main file is a good idea.

File: projects/plugins/jetpack/modules/google-fonts/class-jetpack-google-fonts.php (hunk 3 above)
Comment (jeherve) [reply projects/plugins/jetpack/modules/google-fonts/class-jetpack-google-fonts.php]: I wonder if this may be overkill. When used by third-parties, wouldn't that code pretty much always be used alongside the existing `automattic/jetpack-google-fonts-provider` package? Even if not fully related, maybe it could be a new class inside that package? We could always extract it into a new package if we start needing it elsewhere in the future, somewhere where the jetpack-google-fonts-provider package is not already required.

File: projects/plugins/jetpack/modules/google-fonts/class-jetpack-google-fonts.php [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+		add_action( 'after_setup_theme', array( $this, 'register_google_fonts' ) );
+
+		/**
+		 * We are already enqueueing all registered fonts by default when loading the block editor,
+		 * so we only need to scan for webfonts when browsing as a guest.
+		 */
+		if ( ! is_admin() ) {
+			add_action( 'wp_loaded', array( $this, 'enqueue_global_styles_google_fonts' ) );
+			add_filter( 'pre_render_block', array( $this, 'scan_block_for_google_fonts' ), 10, 2 );
+		}
```
Comment (zaguiini) [inline projects/plugins/jetpack/modules/google-fonts/class-jetpack-google-fonts.php:36]: Need to keep it like that until Gutenberg lazy loads font-face declarations as they're picked in the block editor.

**Comment by jeyip** [comment]
Not entirely certain if it's just me, but I keep running into critical errors when running this locally with the instructions provided.

```
PHP Fatal error:  Uncaught Error: Call to undefined function Automattic\Jetpack\Fonts\introspect_global_styles_webfonts() in /usr/local/src/jetpack-monorepo/projects/plugins/jetpack/modules/google-fonts/class-jetpack-google-fonts.php:144
Stack trace:
#0 /var/www/html/wp-includes/class-wp-hook.php(307): Jetpack_Google_Fonts->enqueue_global_styles_google_fonts()
#1 /var/www/html/wp-includes/class-wp-hook.php(331): WP_Hook->apply_filters()
#2 /var/www/html/wp-includes/plugin.php(474): WP_Hook->do_action()
#3 /var/www/html/wp-settings.php(609): do_action()
#4 /var/www/html/wp-config.php(86): require_once('/var/www/html/w...')
#5 /var/www/html/wp-load.php(50): require_once('/var/www/html/w...')
#6 /var/www/html/wp-blog-header.php(13): require_once('/var/www/html/w...')
#7 /var/www/html/index.php(17): require('/var/www/html/w...')
#8 {main}
```

**Comment by jeyip** [comment]
## Testing
I added EB Garamond to the block settings of a site title block. I also selected Space Mono from the global styles typography panel.

### Requirements
- [x] Only the picked font families show up in the style#webfonts-inline-css element

### Browsers
- [x] Chrome

File: projects/plugins/jetpack/modules/google-fonts/class-jetpack-google-fonts.php (hunk 3 above)
Comment (creativecoder) [reply projects/plugins/jetpack/modules/google-fonts/class-jetpack-google-fonts.php]: > I wonder if this may be overkill.

I would agree; I think separating this out into its own package doesn't really provide much benefit. Realistically, you should always be introspecting fonts when using the provider with more than a few font faces (especially with how Google creates so many font face declarations for different character ranges). I'd hope that packaging the provider and introspection utilities together would encourage that.

If `automattic/jetpack-google-fonts-provider ` as a package is too specific, I'd rather deprecate it and create a new package to more broadly encompass webfont functionality.

Additionally, having Global styles introspection and block introspection in separate places seems odd; I think these should be together.

**Comment by creativecoder** [comment]
> Not entirely certain if it's just me, but I keep running into critical errors when running this locally with the instructions provided.

I was getting the same error, but I think the latest changes (that add `"automattic/jetpack-global-styles-webfonts-introspector": "0.1.x-dev",` to the Jetpack plugin's composer.json) fix the problem.

Note you'll need to run `jetpack install -r ` and/or `jetpack install --all` after pulling.

File: projects/plugins/jetpack/modules/google-fonts/jetpack-google-fonts.php (hunk 5 above)
Comment (creativecoder) [reply projects/plugins/jetpack/modules/google-fonts/jetpack-google-fonts.php:1]: Agreed, having to find your way to the required file with the actual module code seem unnecessary.

File: projects/plugins/jetpack/modules/google-fonts/class-jetpack-google-fonts.php (hunk 3 above)
Comment (zaguiini) [reply projects/plugins/jetpack/modules/google-fonts/class-jetpack-google-fonts.php]: Yep. Any reason why we created this `jetpack-google-fonts-provider` package in the first place? It looks weird to have this as separate parts.

And I do agree -- I think it's better if we collocated the provider and the introspector together within the Jetpack Google Fonts module. Will do that.

Comment (creativecoder) [reply projects/plugins/jetpack/modules/google-fonts/class-jetpack-google-fonts.php]: Sorry if I wasn't clear, I don't mean stop using packages all together. My understanding is that Jetpack (the monorepo) is built around packages for functionality that's reusable across plugins. That way if Woo, for example, wants to use something that's in Jetpack, it doesn't have to copy the code into itself, or require install the entire Jetpack plugin be installed, it can just require the specific package it needs.

So for our use case, I think this separation makes sense:

- Package: font provider, and any functionality needed for loading fonts from the provider. No hooks should be added by default, so it's up to package consumers to "turn on" any functionality from the package they want to use.
- Module: Our WP.com specific list of curated Google Fonts, and calling the necessary functions from the package to load those fonts.

The main goal of the separation here being any other plugin could use the fonts package to load it's own set of Google Fonts, independent from the fonts module in the Jetpack plugin.

Comment (zaguiini) [reply projects/plugins/jetpack/modules/google-fonts/class-jetpack-google-fonts.php]: So, a few things:

- The global styles introspection utils package is not adding any hook -- it's simply getting the global styles and looking for font families. However, that might be incomplete because only global styles introspection is not enough even for this use case as block settings are ignored. One thing I would like to have is an action called `jetpack_content_introspection_webfont_found` or something like that would get fired whenever a font family was found in the content;
- The introspection package is not really related to the provider. Adding the provider does not add introspection, and vice-versa. One could use introspection without necessarily working with Google provider. The provider's job is to handle and emit the enqueued fonts, and that's it. It doesn't care about introspection.
- The module could (and will) use both packages. Other modules might not, and only use the provider, or only use the introspection... I do think they should be decoupled as those are very useful things to expose.

**Comment by creativecoder** [comment]
@zaguiini Here's a draft PR to illustrate what I was trying to say in the comments: https://github.com/Automattic/jetpack/pull/23958

- Hold off on creating a separate package for now, it adds admin overhead, and we can always do this later. If Gutenberg ends up adding this functionality, we'll avoid the hassle of having to deprecate the new package.
- Keep a clean separation between the fonts introspection and fonts module. The module only needs to add the appropriate hooks and the font introspection classes do everything else.
- Keep things modular so it's easy to add another introspector, when font settings are added to other places in Gutenberg.
- (Add some tests to better document how things work.)

Let me know what you think!

**Comment by zaguiini** [comment]
Thanks, @creativecoder!

You made very good points, I agree with all of them.

However, there are some things I'd like to observe...

### All found font faces are enqueued, regardless if we were the ones that registered them

The implementation in my PR tries not to enqueue every single font that gets detected by introspection. [We're keeping track of the fonts we registered](https://github.com/Automattic/jetpack/pull/23810/files#diff-67965f8a19767c6775dd0d0949ed5830c22bda270ec3c1835f801649e98596c2R96-R108), then, when a font is found, [we're checking if it was registered by us before enqueueing](https://github.com/Automattic/jetpack/pull/23810/files#diff-67965f8a19767c6775dd0d0949ed5830c22bda270ec3c1835f801649e98596c2R133-R140).

This tightens things a bit, so only the subset of fonts we registered get enqueued... I wonder how could we implement that intermediary state without the use of a class in a reliable way? One possibility is to get the list of registered fonts from the API (`wp_webfonts()->get_registered_webfonts()`), grab the faces by their slug, and look for the family name of the first font face in the Google Fonts list... But this looks more like a workaround than a solution.

That's why I transitioned into a class, and that's the purpose of the `registered_google_fonts` property in the `Jetpack_Google_Fonts` class: [to have this 1-1 mapping from the slug to the name](https://github.com/Automattic/jetpack/pull/23810/files#diff-67965f8a19767c6775dd0d0949ed5830c22bda270ec3c1835f801649e98596c2R108). Then, we have this `maybe_enqueue_*` method that looks in this map before enqueuing the font family.

Could we achieve something similar without classes? Maybe using global variables or a singleton class but I'm not sure how common this pattern is in Jetpack...

### Introspection classes in the same plugin as the Google fonts provider

It looks really off to have the introspection classes inside the Google fonts provider plugin. Those are different things that could work without each other... But I cannot think of a better place to put it without creating a new plugin. If there's absolutely no better place to put it (how about putting it in the module?), I'm fine with leaving it where it is.

Right now they do have some degree of relationship because as you read in my first point, we want only to enqueue the fonts that we registered.

However, I see them as more general. It would be really nice if we could abstract the global styles and block introspection and apply an action or filter whenever a registered font is found. Exposing that as a plugin, I can see a lot of external plugins using it to do the same thing within their own plugins.

What I'm trying to say is that it's not really tied to Google -- it has a lot of potential reusability by third parties. Unless, of course, we don't want them using it.

### Migrating to Gutenberg

Not specific to our implementations, but how are going to migrate from Jetpack to Gutenberg when it's time? Is there any guideline on how to deal with these situations? Should we look for the Gutenberg implementation before hooking up our own filters? Curious about how it'll work...

---

Sorry for the wall of text, but those are my concerns 😄

File: projects/plugins/jetpack/modules/google-fonts/class-jetpack-google-fonts.php (hunk 3 above)
Comment (jeyip) [reply projects/plugins/jetpack/modules/google-fonts/class-jetpack-google-fonts.php]: > One could use introspection without necessarily working with Google provider. The provider's job is to handle and emit the enqueued fonts, and that's it. It doesn't care about introspection.

Imagining myself as a developer with 0 knowledge of the fonts work we've done, I'd find it unintuitive to need to import two separate packages to enqueue webfonts in a performant way. That is to say, I wouldn't think to look for an introspector while setting up fonts support. 

I agree with Jeremy + Grant -- I suspect there'd be more clarity in colocating both introspection and provider logic in the same package for now.

**Comment by jeyip** [comment]
> Could we achieve something similar without classes? Maybe using global variables or a singleton class but I'm not sure how common this pattern is in Jetpack...

Out of curiosity, is there a reason we'd prefer not to use a class to implement this logic?

File: projects/packages/global-styles-webfonts-introspector/src/class-global-styles-webfonts-introspector.php [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+			if ( isset( $matches['slug'] ) ) {
+				return $matches['slug'];
+			}
+
+			return $font_family;
+		}
+	}
+
+	/**
+	 * Introspect global styles for webfonts.
```
Comment (jeyip) [inline projects/packages/global-styles-webfonts-introspector/src/class-global-styles-webfonts-introspector.php:47]: Could we add an @return inline comment here for consistency?

File: projects/packages/global-styles-webfonts-introspector/src/class-global-styles-webfonts-introspector.php [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+
+			// Full string: var:preset|font-family|slug
+			// We do not care about the origin of the font, only its slug.
+			preg_match( '/font-family\|(?P<slug>.+)$/', $font_family, $matches );
+
+			if ( isset( $matches['slug'] ) ) {
+				return $matches['slug'];
+			}
+
+			return $font_family;
```
Comment (jeyip) [inline projects/packages/global-styles-webfonts-introspector/src/class-global-styles-webfonts-introspector.php:42]: Could you describe what this line would return? Would it just be a font family css variable `var(--wp--preset--font-family--slug)` where a match isn't found?

File: projects/packages/global-styles-webfonts-introspector/src/class-global-styles-webfonts-introspector.php [hunk 11]
```diff
@@ -0,0 +1,90 @@
+<?php
```
Comment (jeyip) [inline projects/packages/global-styles-webfonts-introspector/src/class-global-styles-webfonts-introspector.php:1]: I really appreciate how well structured and easy-to-follow this class is 🙏

File: projects/packages/global-styles-webfonts-introspector/src/class-global-styles-webfonts-introspector.php [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+	public static function introspect_global_styles_webfonts() {
+		if ( ! function_exists( 'gutenberg_get_global_styles' ) ) {
+			return;
+		}
+
+		$global_styles = gutenberg_get_global_styles();
+
+		$found_webfonts = array();
+
+		// Look for fonts in block presets...
```
Comment (jeyip) [inline projects/packages/global-styles-webfonts-introspector/src/class-global-styles-webfonts-introspector.php:58]: `Non-blocking`: How would we feel about providing a little more context for those that might be unfamiliar with global styles?

```
// Global styles can specify typography for individual blocks,
// so look for fonts in block presets...


// Global styles can specify typography for elements like headings, paragraphs, etc.,
// so look for fonts in HTML element presets...
```

File: projects/packages/global-styles-webfonts-introspector/src/class-global-styles-webfonts-introspector.php (hunk 10 above)
Comment (zaguiini) [reply projects/packages/global-styles-webfonts-introspector/src/class-global-styles-webfonts-introspector.php:42]: >Also completely unrelated to this PR, but it seems so odd to have so many different ways of representing the font family slug 🙃

Right! That's because you can specify a CSS variable to link it inside `theme.json`. You can also pick a font in the editor, and then it would become the second format `var:preset|font-family|slug`, which is the one that's registered in the global styles CPT.

But then, you can also solely specify the slug, and that's why we're returning... If none of the formats match, just return what's in there and we'll try our luck. It might be a system font, might be just the slug... Who knows.

File: projects/packages/global-styles-webfonts-introspector/src/class-global-styles-webfonts-introspector.php (hunk 12 above)
Comment (zaguiini) [reply projects/packages/global-styles-webfonts-introspector/src/class-global-styles-webfonts-introspector.php:58]: That's more explicit, for sure! I'm curious if this is necessary since it kind of goes without saying that global styles can specify the typography since we're looking for them in the code?

Comment (jeyip) [reply projects/packages/global-styles-webfonts-introspector/src/class-global-styles-webfonts-introspector.php:58]: > I'm curious if this is necessary since it kind of goes without saying that global styles can specify the typography since we're looking for them in the code?

Sure -- for those examples, I see what you're saying. My intention is to ultimately provide more clarity for others by adding some context about business use cases, especially since this code will live in Jetpack for the time being where other folks might not be as familiar with global styles.

Does that make sense? Maybe we can think about different, more meaningful wording.

File: projects/plugins/jetpack/modules/google-fonts/class-jetpack-google-fonts.php [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+						'font-style'   => 'italic',
+						'font-display' => 'fallback',
+						'provider'     => 'jetpack-google-fonts',
+					),
+				)
+			);
+
+			if ( empty( $font_family_slugs ) ) {
+				// Fonts were not registered.
+				continue;
```
Comment (jeyip) [inline projects/plugins/jetpack/modules/google-fonts/class-jetpack-google-fonts.php:93]: Could this lead to errors being silently hidden?

Comment (zaguiini) [reply projects/plugins/jetpack/modules/google-fonts/class-jetpack-google-fonts.php:93]: Not really -- we're always emitting errors in the front-end so they'll be noticed.

File: projects/packages/global-styles-webfonts-introspector/src/class-global-styles-webfonts-introspector.php (hunk 12 above)
Comment (zaguiini) [reply projects/packages/global-styles-webfonts-introspector/src/class-global-styles-webfonts-introspector.php:58]: Gotcha. Sure, we can think of something.

**Review by jeyip — COMMENTED** [review]
The code generally looks good to me. @creativecoder made a number of great points though

- Regarding modularity and having global styles introspection play well with other introspection logic...from what I can tell, what's written here would work given those considerations.
- Tests! Totally agree.

+1 from me when my last few comments are answered and tests are written.

File: projects/plugins/jetpack/modules/google-fonts/class-jetpack-google-fonts.php (hunk 13 above)
Comment (jeyip) [reply projects/plugins/jetpack/modules/google-fonts/class-jetpack-google-fonts.php:93]: Hmmm...I might be missing something but here's the flow of logic I have in my head ( I might be mistaken about this! )

- We'd attempt to register webfonts
- `register_webfont` method would return `false`
- `wp_register_webfonts` would fail and return an empty array of `font_family_slugs`
- We wouldn't enqueue any fonts because `registered_google_fonts` would be empty
- The frontend would silently fallback to some system or default font

When would errors be emitted on the frontend?

Comment (zaguiini) [reply projects/plugins/jetpack/modules/google-fonts/class-jetpack-google-fonts.php:93]: The `wp_register_webfonts` method runs for every font family declared in the list, so there'd be multiple runs and not a single one.

And yes, that's mostly right. One thing, though, is that the user/developer would be notified after `register_webfont method would return false` on the front-end, so they'd have a clue on how to fix what went wrong.

Comment (jeyip) [reply projects/plugins/jetpack/modules/google-fonts/class-jetpack-google-fonts.php:93]: > register_webfont method would return false on the front-end, so they'd have a clue on how to fix what went wrong

Ah I see -- that's what I was missing. When you have time, would you mind explaining how returning false would notify the user / dev in the frontend? The only way I can see this working is with error logging.

Comment (zaguiini) [reply projects/plugins/jetpack/modules/google-fonts/class-jetpack-google-fonts.php:93]: It's not for the front-end. Return `false` is more about conditional action upon registration failure for example. I do agree, though, that we're lacking descriptive return values on failure when dealing with the back-end (i.e. `false` can mean multiple things). But that's not something to address in this PR.

Comment (jeyip) [reply projects/plugins/jetpack/modules/google-fonts/class-jetpack-google-fonts.php:93]: > But that's not something to address in this PR.

Agreed

Comment (zaguiini) [reply projects/plugins/jetpack/modules/google-fonts/class-jetpack-google-fonts.php:93]: Oh, my bad.

What I tried to say was that a notice is displayed on the front end, so the user can take action.

File: projects/plugins/jetpack/modules/google-fonts/class-jetpack-google-fonts.php [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+		return $content;
+	}
+
+	/**
+	 * Enqueue a font family if it was not removed from the curated list.
+	 *
+	 * @param string $font_family_slug The block content.
+	 */
+	private function maybe_enqueue_font_family( $font_family_slug ) {
+		if ( ! isset( $this->registered_google_fonts[ $font_family_slug ] ) ) {
```
Comment (zaguiini) [inline projects/plugins/jetpack/modules/google-fonts/class-jetpack-google-fonts.php:134]: We need to check if the font is registered before trying to enqueue it:

```suggestion
		if ( ! wp_webfonts()->is_font_family_registered( $font_family_slug ) ) {
```

The `is_font_family_registered` function will be exposed after https://github.com/WordPress/gutenberg/pull/39988 gets merged. For now, what we can use is:

```suggestion
		if ( ! isset( wp_webfonts()->get_registered_webfonts()[ $font_family_slug ] ) ) {
```

**Comment by zaguiini** [comment]
After meeting with @jeyip and @creativecoder, we came up with a few observations:

### We might not need a class for the Jetpack Google fonts module

The reason for having a class was to share the state between hook calls: the one that registers webfonts, and the introspection callbacks (e.g. `pre_render_block`).

This shared state existed so we can be a little more cautious regarding which fonts get enqueued. We wanted to make sure that only the fonts we registered were enqueued, and that's a safety measure. However, that's too cautious.

See, right now, Gutenberg is in this weird state where externally registered webfonts picked in the editor do not have any effect on the front end. Removing [the guard clause](https://github.com/Automattic/jetpack/pull/23810/files#diff-67965f8a19767c6775dd0d0949ed5830c22bda270ec3c1835f801649e98596c2R134-R137) would allow any picked font in the editor to show up in the front end.

This is the worst-case scenario, and one that we actually want to have as it would temporarily fix Gutenberg's behavior (see https://github.com/WordPress/gutenberg/issues/40363 and https://github.com/WordPress/gutenberg/issues/40386).

A note regarding the implementation is this comment: https://github.com/Automattic/jetpack/pull/23810/files#r851520186. We still need to check if the font is registered before enqueueing it since we're not sure it was registered before now that we'll get rid of the intermediary state. Not checking and trying to enqueue a webfont that was not registered will add a notice on the front end.

### We might need additional information on where to put the introspection classes

@jeherve could really help us here as we're uncertain about where to put them. It still feels unnatural to add the introspection classes to the Google provider package. Ideally, we should create a new package to expose a separate thing since one can use the introspection class without necessarily using the Google font provider, and vice-versa.

But, how much of a hassle is it to create, deploy and maintain packages in WPCOM? Should we? How can we balance the trade-off between having a well-organized codebase and the burden of maintaining a new package? The answers are beyond our reach right now, but @Automattic/jetpack-crew might have them all.

### Migrating to Gutenberg

How are going to migrate introspection from Jetpack to Gutenberg when it's time? Is there any guideline on how to deal with these situations? Should we add a guard to the module that would not hook up our own filters in case Gutenberg already did so? Curious about how it'll work...

Introspection is going to happen there at some point (see https://github.com/WordPress/gutenberg/issues/40363 and https://github.com/WordPress/gutenberg/issues/40386), so the introspection code included here is temporary. Please take that into consideration when answering the questions on the "We might need additional information on where to put the introspection classes" paragraph.

### Shepherding the PR to `master`

I'm going OOO for the next three weeks starting on Monday, so @Automattic/cylon will take care of it. Please reach them out as necessary, but I believe the discussion will be centralized here.

**Comment by Addison-Stavlo** [comment]
> How are going to migrate introspection from Jetpack to Gutenberg when it's time?

We don't. We remove the introspection from Jetpack when it we are able.

My takeaways from this long slack thread  (p1649889597371109-slack-CHN6J22MP) with @mtias were:

* Fonts in theme.json and global-styles (saved theme.json) are to be enqueued on the front-end. (p1649957191543249/1649889597.371109-slack-CHN6J22MP)
* Programmatically registered webfonts are not intended to be available anywhere except global styles, the fact that they appear in individual block settings font pickers is a current bug in that system that will need to be resolved (p1649957314019679/1649889597.371109-slack-CHN6J22MP). In the end individual block settings will only list theme supplied fonts and not programmatically registered fonts.

When this is the case, there is no need for all the introspection since fonts available to individual blocks are already enqueued from being part of theme.json and any fonts selected by global styles would be enqueued from there.

The introspection is now only a temporary solution to allow Google fonts on dotcom sooner. This serves to band-aid the gaps in enqueueing that should not be present in the end state of the Gutenberg webfonts API. It fills the gap of not enqueueing global styles selections (which should be resolved soon?). This also allows us to fill the gap in the current/temporary state of registered webfonts showing up in individual block settings - this part seems reasonable to me as once Gutenberg removes these webfonts from this section any previous selections on dotcom should be back-wards compatible due to this introspection code.

**Comment by creativecoder** [comment]
To move this forward, I'd like to get some guidance from @Automattic/ground-control on whether to create a new package or not.

You can see how I've attempted to cleanly divide the Google Fonts provider from the font enqueuing ("introspection") code in this draft PR: https://github.com/Automattic/jetpack/pull/23958.

On the one hand, the enqueuing code is potentially useful as it's own package for use with other font providers, inside or outside Jetpack. On the other, the WP Webfonts API is still maturing and not really stable yet, so significant changes my be needed, or that code might become irrelevant, so it may be worth waiting.

Also, I know creating a separate package incurs some overhead, particularly on wpcom.

Do you have a preference on keeping the code for the Introspectors classes (from #23958) within the Google Fonts provider, or creating a separate package for it, at this time? We'd like to wrap this up and get it shipped asap.

Thanks!

**Comment by jeherve** [comment]
> @jeherve could really help us here as we're uncertain about where to put them. 

Sorry, I was AFK for the past few days, only getting back to this now. 

All in all, and given that this is all building atop something that's not part of WordPress Core, thus not widely adopted yet, I would vote towards keeping it simple so we can move and iterate quickly. In practice, I think that means less packages. Even though a new package may prove useful to third-parties in the future, I don't think it will be useful to many people right now. Since there are still many moving pieces in the new API, I don't think anyone would reasonably start building their own solution on top of 2 separate Jetpack packages that themselves rely on a rapidly changing Gutenberg-only feature.

In short, I would:

- Not worry too much about what we do today; we can always iterate later as things get more stable in Core, and as the questions to Gutenberg are answered.
- Not create a new package today.
- Keep things in the Jetpack plugin itself and in the existing package.
    - `Global_Styles_Webfonts_Introspector` could be a new class in the package.
    - `projects/plugins/jetpack/modules/google-fonts.php` could require and initialize `projects/plugins/jetpack/modules/google-fonts/class-jetpack-google-fonts.php`. We can require more files in `projects/plugins/jetpack/modules/google-fonts.php` if it becomes necessary.

What do you think?

**Comment by creativecoder** [comment]
> Sorry, I was AFK for the past few days, only getting back to this now.

@jeherve No problem! Thanks for prioritizing a reply here so soon after getting back.

> What do you think?

That makes sense to me. I'll close this in favor of https://github.com/Automattic/jetpack/pull/23958 and move forward there.
